packages/flutter_tools/gradle/src/main/kotlin/FlutterPluginUtils.kt

```diff
@@ -371,4 +392,433 @@ object FlutterPluginUtils {
+            if (flutterBuildMode == "release" && (pluginObject["dev_dependency"] as? Boolean == true)) {
+                // This plugin is a dev dependency will not be included in the
+                // release build, so no need to add its dependencies.
+                return@forEach
+            }
+            val dependencies = requireNotNull(pluginObject["dependencies"] as? List<*>)
+            dependencies.forEach innerForEach@{ pluginDependencyName ->
+                check(pluginDependencyName is String)
+                if (pluginDependencyName.isEmpty()) {
+                    return@innerForEach
+                }
+
+                val dependencyProject =
+                    project.rootProject.findProject(":$pluginDependencyName") ?: return@innerForEach
+                pluginProject.afterEvaluate {
+                    pluginProject.dependencies.add("implementation", dependencyProject)
+                }
+            }
+        }
+    }
+
+    @JvmStatic
+    @JvmName("configurePluginProject")
+    internal fun configurePluginProject(
+        project: Project,
+        pluginObject: Map<String?, Any?>,
+        engineVersion: String
+    ) {
+        // TODO(gmackall): should guard this with a pluginObject.contains().
+        val pluginName =
+            requireNotNull(pluginObject["name"] as? String) { "Plugin name must be a string for plugin object: $pluginObject" }
+        val pluginProject: Project = project.rootProject.findProject(":$pluginName") ?: return
+
+        // Apply the "flutter" Gradle extension to plugins so that they can use it's vended
+        // compile/target/min sdk values.
+        pluginProject.extensions.create("flutter", FlutterExtension::class.java)
+
+        // Add plugin dependency to the app project. We only want to add dependency
+        // for dev dependencies in non-release builds.
+        project.afterEvaluate {
+            getAndroidExtension(project).buildTypes.forEach { buildType ->
+                if (!(pluginObject["dev_dependency"] as Boolean) || buildType.name != "release") {
+                    project.dependencies.add("${buildType.name}Api", pluginProject)
+                }
+            }
+        }
+
+        // Wait until the Android plugin loaded.
+        pluginProject.afterEvaluate {
+            // Checks if there is a mismatch between the plugin compileSdkVersion and the project compileSdkVersion.
+            val projectCompileSdkVersion: String = getCompileSdkFromProject(project)
+            val pluginCompileSdkVersion: String = getCompileSdkFromProject(pluginProject)
+            // TODO(gmackall): This is doing a string comparison, which is odd and also can be wrong
+            //                 when comparing preview versions (against non preview, and also in the
+            //                 case of alphabet reset which happened with "Baklava".
+            if (pluginCompileSdkVersion > projectCompileSdkVersion) {
```

Comment:
Consider breaking this out into its own method so that when we fix the comparison the unit tests are easy to write/modify.